### PR TITLE
build(tests): type check packages/neovim/*.test.ts

### PR DESCRIPTION
**Pull request by gjf7**
Follow up #410, for #352.

### CODE REVIEW

File: packages/neovim/tsconfig.json [hunk 1]
```diff
@@ -3,6 +3,6 @@
   "compilerOptions": {
     "outDir": "./lib"
   },
-  "include": ["src", "**/api/Buffer.test.ts"],
-  "exclude": ["node_modules", "__tests__", "**/*.test.ts", "examples", "lib"]
+  "include": ["src"],
+  "exclude": ["node_modules", "examples", "lib"]
```
Comment (justinmk) [inline packages/neovim/tsconfig.json:7]: Nice, so we can close https://github.com/neovim/node-client/issues/352 ?

File: packages/neovim/src/api/Window.test.ts [hunk 2]
```diff
@@ -105,15 +107,15 @@ describe('Window API', () => {
 
       windows = await nvim.windows;
       // Set to new split
-      [, nvim.window] = windows;
+      nvim.window = windows[1] as unknown as AsyncWindow;
```
Comment (justinmk) [inline packages/neovim/src/api/Window.test.ts:110]: can this be avoided by changing the `let windows` delcaration on  line 99 to:

      let windows: Awaited<typeof nvim.windows>;

File: packages/neovim/src/api/Neovim.test.ts [hunk 3]
```diff
@@ -24,8 +28,8 @@ describe('Neovim API', () => {
       expect(await nvim.buffer.name).toMatch(/goodbye\.txt$/);
 
       // switch buffers
-      [nvim.buffer] = buffers;
-      expect(await nvim.buffer.name).toMatch(/hello\.txt$/);
+      nvim.buffer = buffers[0] as unknown as AsyncBuffer;
```
Comment (justinmk) [inline packages/neovim/src/api/Neovim.test.ts:31]:
```diff
diff --git a/packages/neovim/src/api/Neovim.ts b/packages/neovim/src/api/Neovim.ts
index 5929a6b5a57d..6a7c5c6e7eef 100644
--- a/packages/neovim/src/api/Neovim.ts
+++ b/packages/neovim/src/api/Neovim.ts
@@ -169,7 +169,7 @@ export class Neovim extends BaseApi {
   /**
    * Sets the current buffer
    */
-  set buffer(buffer: AsyncBuffer) {
+  set buffer(buffer: AsyncBuffer | Buffer) {
     this.request(`${this.prefix}set_current_buf`, [buffer]);
   }
```

File: packages/neovim/tsconfig.json (hunk 1 above)
Comment (gjf7) [reply packages/neovim/tsconfig.json:7]: Not really, still need to work on packages/integration-tests/.

File: packages/neovim/src/api/Window.test.ts (hunk 2 above)
Comment (gjf7) [reply packages/neovim/src/api/Window.test.ts]: That's look way better!

File: packages/neovim/src/api/Tabpage.test.ts [hunk 4]
```diff
@@ -41,7 +42,7 @@ describe('Tabpage API', () => {
       const tabpages = await nvim.tabpages;
       expect(tabpages.length).toBe(2);
 
-      nvim.tabpage = tabpages[tabpages.length - 1];
+      nvim.tabpage = tabpages[tabpages.length - 1] as unknown as AsyncTabpage;
```
[the remark this answers is not in the document]
Comment (gjf7) [reply packages/neovim/src/api/Tabpage.test.ts]: Fixed

File: packages/neovim/src/api/Neovim.test.ts [hunk 5]
```diff
@@ -218,7 +222,7 @@ describe('Neovim API', () => {
       const numWindows = (await nvim.windows).length;
       const buffer = await nvim.createBuffer(false, false);
       expect(await nvim.buffers).toHaveLength(numBuffers + 1);
-
+      assert(buffer instanceof Buffer);
```
[the remark this answers is not in the document]
Comment (gjf7) [reply packages/neovim/src/api/Neovim.test.ts]: It's ok on my mac and ci. Anyway, I changed this to `assert(typeof buffer !== 'number')`.

File: packages/neovim/src/api/Neovim.ts [hunk 6]
```diff
@@ -232,7 +232,7 @@ export class Neovim extends BaseApi {
   /**
    * Sets the current tabpage
    */
-  set tabpage(tabpage: AsyncTabpage) {
+  set tabpage(tabpage: AsyncTabpage | Tabpage) {
     this.request(`${this.prefix}set_current_tabpage`, [tabpage]);
```
Comment (justinmk) [inline packages/neovim/src/api/Neovim.ts:236]: I wonder if this and the buffer() case were a hidden bug. Do they need the same treatment as in window()? I don't see anywhere in `this.request` codepath that would implicitly handle a Promise (AsyncBuffer/AsyncTabpage)

```ts
    if (tabpage instanceof Tabpage) this.request(`${this.prefix}set_current_tabpage`, [tabpage]);
    else tabpage.then(tabpage => this.request(`${this.prefix}set_current_tabpage`, [tabpage]));
```

**Review by justinmk — APPROVED** [review]

**Comment by justinmk** [comment]
https://github.com/neovim/node-client/pull/411#discussion_r1795432889 doesn't need to block this, but might need a fix.

**Pull request merged**